arbstate/daprovider/reader.go

```diff
@@ -0,0 +1,222 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package daprovider
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/das/dastree"
+	"github.com/offchainlabs/nitro/util/blobs"
+)
+
+type Reader interface {
+	// IsValidHeaderByte returns true if the given headerByte has bits corresponding to the DA provider
+	IsValidHeaderByte(headerByte byte) bool
+
+	// RecoverPayloadFromBatch fetches the underlying payload from the DA provider given the batch header information
+	RecoverPayloadFromBatch(
+		ctx context.Context,
+		batchNum uint64,
+		batchBlockHash common.Hash,
+		sequencerMsg []byte,
+		preimageRecorder PreimageRecorder,
+		validateSeqMsg bool,
+	) ([]byte, error)
+
+	// RecordPreimagesTo takes in preimages map and returns a function that can be used
+	// In recording (hash,preimage) key value pairs into preimages map, when fetching payload
+	RecordPreimagesTo(
+		preimages map[arbutil.PreimageType]map[common.Hash][]byte,
+	) PreimageRecorder
+}
+
+type PreimageRecorder func(key common.Hash, value []byte)
```

Comment:
I like PreimageRecorder, it's much clearer now.
I would change it's type to also get primageType parameter, then RecordPreimagesTo could be a public function outside the interface.
dastree.RecordHash could be changed accordingly, e.g. to also accept this function type and always use keccak.

Reply:
(this will also allow DAS providers to mix and match preimage types if they feel like it)

Reply:
Agreed!